crates/risc0/src/host/boundless.rs

```diff
@@ -0,0 +1,710 @@
+use tracing::Instrument;
+use url::Url;
+use uuid::Uuid;
+
+use super::config::{get_boundless_builtin_storage_provider, BoundlessConfig};
+use crate::host::pricing_service::{PriceResponse, PricingService};
+
+enum ResubmitResult {
+    Retry,
+    Success,
+}
+
+#[derive(Clone)]
+pub struct BoundlessProver {
+    pub client: Client,
+    pub ledger_db: LedgerDB,
+    pub pricing_service: PricingService,
+}
+
+impl BoundlessProver {
+    pub async fn new(ledger_db: LedgerDB) -> Self {
+        assert!(
+            std::env::var("RISC0_PROVER").is_ok_and(|prover| prover == "boundless"),
+            "RISC0_PROVER must be explicitly set to boundless"
+        );
+        let client = Self::boundless_client()
+            .await
+            .expect("Failed to create boundless client");
+
+        assert!(
+            client.storage_provider.is_some(),
+            "a storage provider is required to upload the zkVM guest ELF"
+        );
+        Self {
+            client,
+            ledger_db,
+            pricing_service: PricingService::new(),
+        }
+    }
+
+    async fn boundless_client() -> anyhow::Result<Client> {
+        let config = BoundlessConfig::from_env().expect("Failed to load boundless config");
+
+        // If in dev mode, uses a temporary file as storage provider
+        // Otherwise first tries to parse pinata env variables
+        // If fails then tries to parse s3 env variables
+        // If the environment variable `RISC0_DEV_MODE` is set, a temporary file storage provider is used.
+        // Otherwise, the environment variables in `BoundlessPinataStorageConfig` or `BoundlessS3StorageConfig` is checked
+        let storage_provider = get_boundless_builtin_storage_provider().await?;
+
+        // Create a Boundless client from the provided parameters.
+        ClientBuilder::new()
+            .with_deployment(config.deployment)
+            .with_rpc_url(config.rpc_url)
+            .with_storage_provider(Some(storage_provider))
+            .with_private_key(config.wallet_private_key)
+            .build()
+            .await
+    }
+
+    pub async fn prove(
+        &self,
+        job_id: Uuid,
+        elf: Vec<u8>,
+        input: Vec<u8>,
+        assumptions: Vec<AssumptionReceipt>,
+        receipt_type: ReceiptType,
+    ) -> anyhow::Result<oneshot::Receiver<ProofWithJob>> {
+        // Upload image id
+        let image_id = compute_image_id(&elf).expect("Invalid elf program");
+
+        assert!(
+            !is_dev_mode(),
+            "RISC0_DEV_MODE should not be set for boundless"
+        );
+
+        assert!(
+            matches!(receipt_type, ReceiptType::Groth16),
+            "Currently, only Groth16 receipts are supported for boundless"
+        );
+
+        // TODO: Can be done better?
+        let s3_url = read_env("BOUNDLESS_S3_URL")?;
+
+        let s3_use_presigned = read_env("BOUNDLESS_S3_NO_PRESIGNED")
+            .map(|_| false)
+            .unwrap_or(true);
```

Comment:
```suggestion
        let s3_use_presigned = read_env("BOUNDLESS_S3_NO_PRESIGNED").is_err()
```

Or even better, switching to `BOUNDLESS_S3_USE_PRESIGNED` and use is_ok() instead

Reply:
Only used it this confusing way to make it same with boundless